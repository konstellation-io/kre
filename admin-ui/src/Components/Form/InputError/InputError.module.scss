@import 'Styles/mixins';
@import 'Styles/colors';

.errorMessage {
  @include input-footer($color-error);
}
