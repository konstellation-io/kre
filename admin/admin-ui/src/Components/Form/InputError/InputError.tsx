import React from 'react';
import cx from 'classnames';
import styles from './InputError.module.scss';

type Props = {
  message?: string;
};

function InputError({ message = '' }: Props) {
  return (
    <div
      className={cx(styles.errorMessage, { [styles.show]: message !== '' })}
      data-testid="error-message"
    >
      {message}
    </div>
  );
}

export default InputError;
