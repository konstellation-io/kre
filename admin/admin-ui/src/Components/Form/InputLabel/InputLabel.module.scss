@import 'Styles/mixins';
@import 'Styles/colors';

.label {
  @include font-caption;

  display: block;
  letter-spacing: 0.4px;
  color: font-color(grey);
  margin-bottom: 11px;
  min-height: 13px;

  transition: color ease 0.3s;

  &.hidden {
    color: transparent;
  }
}
