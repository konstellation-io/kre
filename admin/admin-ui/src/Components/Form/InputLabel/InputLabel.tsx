import React from 'react';
import cx from 'classnames';
import styles from './InputLabel.module.scss';

type Props = {
  text?: string;
  hidden?: boolean;
};

function InputLabel({ text = '', hidden = false }: Props) {
  return (
    <label
      className={cx(styles.label, { [styles.hidden]: hidden })}
      data-testid="label"
    >
      {text.toUpperCase()}
    </label>
  );
}

export default InputLabel;
