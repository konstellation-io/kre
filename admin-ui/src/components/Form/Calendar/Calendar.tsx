import React, { useState, useEffect } from 'react';
import moment, { Moment } from 'moment';
import { isInclusivelyBeforeDay } from 'react-dates';

import InputLabel from '../InputLabel/InputLabel';
import InputError from '../InputError/InputError';
import ArrowRightIcon from '@material-ui/icons/ArrowForward';
import ArrowLeftIcon from '@material-ui/icons/ArrowBack';
import { DateRangePicker } from 'react-dates';
import CalendarTimeControls from './components/CalendarTimeControls';

type TimeUnit = 'day' | 'week' | 'month' | 'year';
export type Range = {
  label: string;
  startDate: Moment;
  endDate: Moment;
};
type GenerateActRangeParams = {
  unit: TimeUnit;
  date?: Moment;
  label?: string;
  prefix?: string;
};
function generateRange({
  unit,
  date = moment(),
  label = '',
  prefix = 'This'
}: GenerateActRangeParams): Range {
  return {
    label: label || `${prefix} ${unit}`,
    startDate: date.clone().startOf(unit),
    endDate: date.clone().endOf(unit)
  };
}
function generatePrevRange(unit: TimeUnit, label?: string): Range {
  return generateRange({
    date: moment().subtract(1, unit),
    prefix: 'Previous',
    label,
    unit
  });
}

// Generate quick ranges for calendar extension
function getRanges(): Range[] {
  return [
    generateRange({ unit: 'day', label: 'Today' }),
    generateRange({ unit: 'week' }),
    generateRange({ unit: 'month' }),
    generateRange({ unit: 'year' }),
    generatePrevRange('day', 'Yesterday'),
    generatePrevRange('week'),
    generatePrevRange('month'),
    generatePrevRange('year'),
    {
      label: 'Last 2 days',
      startDate: moment()
        .subtract(1, 'day')
        .startOf('day'),
      endDate: moment().endOf('day')
    },
    {
      label: 'Last 6 months',
      startDate: moment()
        .subtract(6, 'month')
        .startOf('day'),
      endDate: moment().endOf('month')
    }
  ];
}

type DateKey = 'startDate' | 'endDate' | null;
type Dates = {
  startDate: Moment | null;
  endDate: Moment | null;
};
type Props = {
  label?: string;
  error?: string | boolean;
  formFromDate?: Moment | null;
  formToDate?: Moment | null;
  onChangeFromDateInput?: Function;
  onChangeToDateInput?: Function;
  hideError?: boolean;
  addTimeControls?: boolean;
  keepOpen?: boolean;
  autoFocus?: boolean;
};

function Calendar({
  label = '',
  error = '',
  hideError = false,
  formFromDate = null,
  formToDate = null,
  onChangeFromDateInput = function() {},
  onChangeToDateInput = function() {},
  addTimeControls = false,
  autoFocus = false,
  keepOpen = false
}: Props) {
  const [fromDate, setFromDate] = useState<Moment | null>(null);
  const [toDate, setToDate] = useState<Moment | null>(null);
  const [focusedInput, setFocusedInput] = useState<DateKey>(
    autoFocus ? 'startDate' : null
  );

  useEffect(() => {
    setFromDate(formFromDate);
  }, [formFromDate]);

  useEffect(() => {
    setToDate(formToDate);
  }, [formToDate]);

  function changeFromDate(date: Moment | null) {
    setFromDate(date);
    onChangeFromDateInput(date);
  }

  function changeToDate(date: Moment | null) {
    setToDate(date);
    onChangeToDateInput(date);
  }

  function getTimeUnits(date: Moment | null) {
    return {
      hour: date?.get('hour'),
      minute: date?.get('minute'),
      second: date?.get('second')
    };
  }

  function initializeTime(date: Moment, isEnd = false) {
    return isEnd ? date.endOf('day') : date.startOf('day');
  }

  function preserveTime(newDate: Moment, previousDate: Moment) {
    return newDate.set(getTimeUnits(previousDate));
  }

  function setTime(
    previousDate: Moment | null,
    newDate: Moment | null,
    isEnd = false
  ): Moment | null {
    if (newDate === null) return null;

    const newDateWithTime =
      previousDate === null
        ? initializeTime(newDate, isEnd)
        : preserveTime(newDate, previousDate);

    return newDateWithTime;
  }

  return (
    <div>
      {label && <InputLabel text={label} />}
      <DateRangePicker
        startDate={fromDate}
        startDateId="calendar_from_date"
        startDatePlaceholderText="From Date"
        endDate={toDate}
        endDateId="calendar_to_date"
        endDatePlaceholderText="To Date"
        displayFormat={addTimeControls ? 'MM/DD/YYYY HH:mm' : 'MM/DD/YYYY'}
        onDatesChange={({
          startDate: newFromDate,
          endDate: newToDate
        }: Dates) => {
          const newFromDatetime = setTime(fromDate, newFromDate);
          const newToDatetime = setTime(toDate, newToDate, true);

          changeFromDate(newFromDatetime);
          changeToDate(newToDatetime);

          setFocusedInput('startDate');
        }}
        keepOpenOnDateSelect={addTimeControls}
        focusedInput={focusedInput}
        onFocusChange={(input: DateKey) => {
          if (!input && keepOpen) return;

          setFocusedInput(input);
        }}
        customArrowIcon={<ArrowRightIcon className="icon-small" />}
        navPrev={
          <div className="DayPickerNavigation_button DayPickerNavigation_button_1 DayPickerNavigation_button__default DayPickerNavigation_button__default_2 DayPickerNavigation_button__horizontal DayPickerNavigation_button__horizontal_3 DayPickerNavigation_button__horizontalDefault DayPickerNavigation_button__horizontalDefault_4 DayPickerNavigation_leftButton__horizontalDefault DayPickerNavigation_leftButton__horizontalDefault_5">
            <ArrowLeftIcon className="icon-small" style={{ margin: '0 8px' }} />
          </div>
        }
        navNext={
          <div className="DayPickerNavigation_button DayPickerNavigation_button_1 DayPickerNavigation_button__default DayPickerNavigation_button__default_2 DayPickerNavigation_button__horizontal DayPickerNavigation_button__horizontal_3 DayPickerNavigation_button__horizontalDefault DayPickerNavigation_button__horizontalDefault_4 DayPickerNavigation_rightButton__horizontalDefault DayPickerNavigation_rightButton__horizontalDefault_5">
            <ArrowRightIcon
              className="icon-small"
              style={{ margin: '0 8px' }}
            />
          </div>
        }
        isOutsideRange={day => !isInclusivelyBeforeDay(day, moment())}
        minimumNights={0}
        renderCalendarInfo={() =>
          addTimeControls ? (
            <CalendarTimeControls
              ranges={getRanges()}
              changeFromDate={changeFromDate}
              changeToDate={changeToDate}
              formFromDate={fromDate}
              formToDate={toDate}
            />
          ) : (
            <></>
          )
        }
      />
      {hideError === false && <InputError message={error.toString()} />}
    </div>
  );
}

export default Calendar;
