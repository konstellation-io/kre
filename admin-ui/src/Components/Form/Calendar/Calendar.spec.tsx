import React from 'react';
import Calendar from './Calendar';
import { shallow } from 'enzyme';
import InputLabel from '../InputLabel/InputLabel';
import InputError from '../InputError/InputError';
import { DateRangePicker } from 'react-dates';
import moment from 'moment';

describe('Calendar', () => {
  let wrapper: any;

  const onChangeFromDateMock = jest.fn();
  const onChangeToDateMock = jest.fn();

  beforeEach(() => {
    wrapper = shallow(
      <Calendar
        label="calendar"
        hideError
        onChangeFromDateInput={onChangeFromDateMock}
        onChangeToDateInput={onChangeToDateMock}
      />
    );
  });

  it('matches snapshots', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('shows right components', () => {
    expect(wrapper.exists(DateRangePicker)).toBeTruthy();
    expect(wrapper.exists(InputLabel)).toBeTruthy();
    expect(wrapper.exists(InputError)).toBeFalsy();
  });

  it('shows error when any', () => {
    wrapper.setProps({ hideError: false });
    wrapper.setProps({ error: 'some error' });
    expect(wrapper.exists(InputError)).toBeTruthy();
    expect(wrapper.find(InputError).prop('message')).toBe('some error');
  });

  it('should call onChangeProp when DateRangePicker dates changes', () => {
    // Arrange.
    const startMock = moment();
    const endMock = moment();
    // Act.
    wrapper
      .find(DateRangePicker)
      .simulate('datesChange', { startDate: startMock, endDate: endMock });

    // Assert.
    expect(onChangeFromDateMock).toBeCalledWith(startMock);
    expect(onChangeToDateMock).toBeCalledWith(endMock);
  });
  // TODO: Check why Icons are changed to UNDEFINED in snapshots
});
