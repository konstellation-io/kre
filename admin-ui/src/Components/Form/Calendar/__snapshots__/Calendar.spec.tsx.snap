// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar matches snapshots 1`] = `
<div>
  <InputLabel
    text="calendar"
  />
  <withStyles(DateRangePicker)
    anchorDirection="left"
    appendToBody={false}
    block={false}
    calendarInfoPosition="bottom"
    customArrowIcon={
      <UNDEFINED
        className="icon-small"
      />
    }
    customCloseIcon={null}
    customInputIcon={null}
    dayPickerNavigationInlineStyles={null}
    daySize={39}
    disableScroll={false}
    disabled={false}
    displayFormat="MM/DD/YYYY"
    enableOutsideDays={false}
    endDate={null}
    endDateId="calendar_to_date"
    endDatePlaceholderText="To Date"
    firstDayOfWeek={null}
    focusedInput={null}
    hideKeyboardShortcutsPanel={false}
    horizontalMargin={0}
    initialVisibleMonth={null}
    inputIconPosition="before"
    isDayBlocked={[Function]}
    isDayHighlighted={[Function]}
    isOutsideRange={[Function]}
    isRTL={false}
    keepFocusOnInput={false}
    keepOpenOnDateSelect={false}
    minimumNights={0}
    monthFormat="MMMM YYYY"
    navNext={
      <div
        className="DayPickerNavigation_button DayPickerNavigation_button_1 DayPickerNavigation_button__default DayPickerNavigation_button__default_2 DayPickerNavigation_button__horizontal DayPickerNavigation_button__horizontal_3 DayPickerNavigation_button__horizontalDefault DayPickerNavigation_button__horizontalDefault_4 DayPickerNavigation_rightButton__horizontalDefault DayPickerNavigation_rightButton__horizontalDefault_5"
      >
        <UNDEFINED
          className="icon-small"
          style={
            Object {
              "margin": "0 8px",
            }
          }
        />
      </div>
    }
    navPosition="navPositionTop"
    navPrev={
      <div
        className="DayPickerNavigation_button DayPickerNavigation_button_1 DayPickerNavigation_button__default DayPickerNavigation_button__default_2 DayPickerNavigation_button__horizontal DayPickerNavigation_button__horizontal_3 DayPickerNavigation_button__horizontalDefault DayPickerNavigation_button__horizontalDefault_4 DayPickerNavigation_leftButton__horizontalDefault DayPickerNavigation_leftButton__horizontalDefault_5"
      >
        <UNDEFINED
          className="icon-small"
          style={
            Object {
              "margin": "0 8px",
            }
          }
        />
      </div>
    }
    noBorder={false}
    numberOfMonths={2}
    onClose={[Function]}
    onDatesChange={[Function]}
    onFocusChange={[Function]}
    onNextMonthClick={[Function]}
    onPrevMonthClick={[Function]}
    openDirection="down"
    orientation="horizontal"
    phrases={
      Object {
        "calendarLabel": "Calendar",
        "chooseAvailableEndDate": [Function],
        "chooseAvailableStartDate": [Function],
        "clearDates": "Clear Dates",
        "closeDatePicker": "Close",
        "dateIsSelected": [Function],
        "dateIsSelectedAsEndDate": [Function],
        "dateIsSelectedAsStartDate": [Function],
        "dateIsUnavailable": [Function],
        "enterKey": "Enter key",
        "escape": "Escape key",
        "focusStartDate": "Interact with the calendar and add the check-in date for your trip.",
        "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
        "homeEnd": "Home and end keys",
        "jumpToNextMonth": "Move forward to switch to the next month.",
        "jumpToPrevMonth": "Move backward to switch to the previous month.",
        "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardShortcuts": "Keyboard Shortcuts",
        "leftArrowRightArrow": "Right and left arrow keys",
        "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
        "moveFocusByOneMonth": "Switch months.",
        "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
        "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
        "openThisPanel": "Open this panel.",
        "pageUpPageDown": "page up and page down keys",
        "questionMark": "Question mark",
        "returnFocusToInput": "Return to the date input field.",
        "roleDescription": "datepicker",
        "selectFocusedDate": "Select the date in focus.",
        "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
        "upArrowDownArrow": "up and down arrow keys",
      }
    }
    readOnly={false}
    regular={false}
    renderCalendarInfo={[Function]}
    renderDayContents={null}
    renderMonthElement={null}
    renderMonthText={null}
    renderNavNextButton={null}
    renderNavPrevButton={null}
    renderWeekHeaderElement={null}
    reopenPickerOnClearDates={false}
    required={false}
    screenReaderInputMessage=""
    showClearDates={false}
    showDefaultInputIcon={false}
    small={false}
    startDate={null}
    startDateId="calendar_from_date"
    startDatePlaceholderText="From Date"
    verticalHeight={null}
    verticalSpacing={22}
    weekDayFormat="dd"
    withFullScreenPortal={false}
    withPortal={false}
  />
</div>
`;
